velox/dwio/dwrf/reader/SelectiveTimestampColumnReader.cpp

```diff
@@ -98,64 +114,128 @@ void SelectiveTimestampColumnReader::readHelper(RowSet rows) {
-    vector_size_t offset,
-    RowSet rows,
-    const uint64_t* incomingNulls) {
-  prepareRead<int64_t>(offset, rows, incomingNulls);
-  VELOX_CHECK(!scanSpec_->filter());
-  bool isDense = rows.back() == rows.size() - 1;
-  if (isDense) {
-    readHelper<true>(rows);
-  } else {
-    readHelper<false>(rows);
-  }
-  readOffset_ += rows.back() + 1;
-}
+  // Merge the seconds and nanos into 'values_'
+  auto secondsData = secondsValues_->as<int64_t>();
+  auto nanosData = values_->as<uint64_t>();
+  const auto rawNulls = nullsInReadRange_
+      ? (isDense ? nullsInReadRange_->as<uint64_t>() : rawResultNulls_)
+      : nullptr;
+  auto tsValues = AlignedBuffer::allocate<Timestamp>(numValues_, &memoryPool_);
+  auto rawTs = tsValues->asMutable<Timestamp>();
 
-namespace {
-void fillTimestamps(
-    Timestamp* timestamps,
-    const uint64_t* nullsPtr,
-    const int64_t* secondsPtr,
-    const uint64_t* nanosPtr,
-    vector_size_t numValues) {
-  for (vector_size_t i = 0; i < numValues; i++) {
-    if (!nullsPtr || !bits::isBitNull(nullsPtr, i)) {
-      auto nanos = nanosPtr[i];
+  for (vector_size_t i = 0; i < numValues_; i++) {
+    if (!rawNulls || !bits::isBitNull(rawNulls, i)) {
+      auto nanos = nanosData[i];
       uint64_t zeros = nanos & 0x7;
       nanos >>= 3;
       if (zeros != 0) {
         for (uint64_t j = 0; j <= zeros; ++j) {
           nanos *= 10;
         }
       }
-      auto seconds = secondsPtr[i] + EPOCH_OFFSET;
+      auto seconds = secondsData[i] + EPOCH_OFFSET;
       if (seconds < 0 && nanos != 0) {
         seconds -= 1;
       }
-      timestamps[i] = Timestamp(seconds, nanos);
+      rawTs[i] = Timestamp(seconds, nanos);
+    }
+  }
+  values_ = tsValues;
+  rawValues_ = values_->asMutable<char>();
+
+  // Treat the filter as kAlwaysTrue if any of the following conditions are met:
+  // 1) No filter found;
+  // 2) Filter is kIsNotNull but rawNulls==NULL (no elements is null).
+  switch (
+      !filter || (filter->kind() == common::FilterKind::kIsNotNull && !rawNulls)
+          ? common::FilterKind::kAlwaysTrue
+          : filter->kind()) {
+    case common::FilterKind::kAlwaysTrue:
+      // Simply add all rows to output.
+      for (vector_size_t i = 0; i < numValues_; i++) {
+        addOutputRow(rows[i]);
+      }
+      break;
+    case common::FilterKind::kIsNull:
+      processNulls(true, rows, rawNulls);
+      break;
+    case common::FilterKind::kIsNotNull:
+      processNulls(false, rows, rawNulls);
+      break;
+    case common::FilterKind::kTimestampRange:
+    case common::FilterKind::kMultiRange:
+      processFilter(filter, rows, rawNulls);
+      break;
+    default:
+      VELOX_UNSUPPORTED("Unsupported filter.");
+  }
+}
+
+void SelectiveTimestampColumnReader::processNulls(
+    const bool isNull,
+    const RowSet rows,
+    const uint64_t* rawNulls) {
+  if (!rawNulls) {
+    return;
+  }
+  auto rawTs = values_->asMutable<Timestamp>();
+
+  returnReaderNulls_ = false;
```

Comment:
Do we really need to fix this value here?  I think it should be set to the correct value when first set, no?

Reply:
I think so. `TEST_F(E2EFilterTest, timestamp)` will fail as it tries to create a `Timestamp` which is `null` .

Reply:
When `returnReaderNulls_` is true, we probably don't have `rawResultNulls_` allocated and initialized, how do we handle that?

Reply:
I haven't seen it happen yet. Could you elaborate how to test such a case?

Reply:
OK I see if we have a filter on this column then this will never be true.